### PR TITLE
Name service ports based on protocol

### LINKED ISSUE #2069: Follow istio's service port requirements
Right now a service port looks like this:
```yaml
  ports:
  - port: 5601
    protocol: TCP
    targetPort: 5601
```
but istio requires to set a name so that it can know about the protocol used on top of TCP:
https://istio.io/docs/setup/additional-setup/requirements/

so, if TLS is used, it should look like that:
```yaml
  ports:
  - name: https
    port: 5601
    protocol: TCP
    targetPort: 5601
```
or if TLS is disabled, like that:
```yaml
  ports:
  - name: http
    port: 5601
    protocol: TCP
    targetPort: 5601
```

same with the es service.

### PR DESCRIPTION
**Pull request by charith-elastic**
Istio relies on service port names being prefixed with their protocol: https://istio.io/docs/setup/additional-setup/requirements/

This PR names the service port of Elasticsearch, Kibana or APM Server as `http` or `https` based on the configuration supplied in the `http` section of the manifest.

Related to #2064 
Fixes #2069

### CODE REVIEW

File: pkg/controller/apmserver/services_test.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+				return svc
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			apm := mkAPMServer(tc.httpConf)
+			haveSvc := NewService(apm)
+			compare.JSONEqual(t, tc.wantSvc(), haveSvc)
```
Comment (anyasabo) [inline pkg/controller/apmserver/services_test.go]: what does comparing marshalled json get us that comparing structs does not?

File: pkg/utils/compare/json.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// JSONEqual compares the JSON representation of two objects to ensure they are equal.
+func JSONEqual(t *testing.T, want, have interface{}) {
+	t.Helper()
```
Comment (sebgl) [inline pkg/utils/compare/json.go]: TIL this exists

File: pkg/controller/apmserver/services_test.go (hunk 1 above)
Comment (sebgl) [reply pkg/controller/apmserver/services_test.go]: I guess we have several options:
- `reflect.DeepEqual()`
- string representation comparison through json marshalling (what this PR does)
- `hash.Object()` comparison (that, behind the scenes, relies on a string representation of the object through spew)
- https://github.com/google/go-cmp

In unit tests I don't think it matters much, even though performance may vary 🤷‍♂  (I would personally have used `reflect.DeepEqual()` since builtin but it may have some corner cases I don't know about?)

File: pkg/controller/elasticsearch/services/services_test.go [hunk 3]
```diff
@@ -152,3 +156,103 @@ func TestElasticsearchURL(t *testing.T) {
...
+			httpConf: commonv1beta1.HTTPConfig{
+				TLS: commonv1beta1.TLSOptions{
+					Certificate: commonv1beta1.SecretRef{
+						SecretName: "my-cert",
+					},
+				},
+			},
+			wantSvc: func() corev1.Service {
+				svc := mkService()
+				svc.Spec.Ports[0].Name = "https"
```
Comment (sebgl) [inline pkg/controller/elasticsearch/services/services_test.go]: nit: would passing the name as argument help a little bit in reading the tests? `mkService("https")`

File: pkg/controller/apmserver/services_test.go (hunk 1 above)
Comment (charith-elastic) [reply pkg/controller/apmserver/services_test.go]: The main reason is that `reflect.DeepEqual` and `require.Equal` produce diffs that are very difficult to read when there are lots of nested elements. JSON diff is much nicer because the output actually shows you where the problem is. Since all three require reflection (the heaviest part) I think the tiny overhead of JSON conversion is acceptable in a unit testing scenario.

I would love to use `go-cmp` but decided against adding an extra dependency for just this test.

File: pkg/controller/elasticsearch/services/services_test.go (hunk 3 above)
Comment (charith-elastic) [reply pkg/controller/elasticsearch/services/services_test.go]: I kept the function generic because there are a few more tests that I am planning to add in a future PR.

File: pkg/controller/apmserver/services_test.go (hunk 1 above)
Comment (sebgl) [reply pkg/controller/apmserver/services_test.go]: Thanks for the details 👍 
We also use "github.com/go-test/deep" in other parts of the code. Do you know how does it compare to the json conversion? I feel like we should settle on one and use it everywhere.

**Comment by PaulGrandperrin** [comment]
Nice!
Could this be ported to the `1.0` branch or is it too big / too late?

**Comment by charith-elastic** [comment]
This will be included in the next release -- which will be a 1.x release as well. The schedule is yet to be determined though.

In the mean time, you can override the port name in the manifest as follows:

```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1beta1
kind: Elasticsearch
metadata:
  name: elasticsearch-istio
spec:
  version: 7.4.0
  http:
    service:
      spec:
        ports:
          - name: http
            port: 9200
            targetPort: 9200
```

The configuration is similar for Kibana and APM Server as well: https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-common-k8s-elastic-co-v1beta1.html#common-k8s-elastic-co-v1beta1-httpconfig
